imagetest/testworkflow.go

```diff
@@ -0,0 +1,245 @@
+		copyGCSObject.Destination = ts.destination + "/outs"
+		copyGCSObjects := &daisy.CopyGCSObjects{copyGCSObject}
+		copyStep, err := ts.wf.NewStep("copy-objects")
+		if err != nil {
+			return fmt.Errorf("failed to add copy-objects step to workflow %s: %v", ts.Name, err)
+		}
+		copyStep.CopyGCSObjects = copyGCSObjects
+
+		// The "copy-objects" step depends on every wait step.
+		for stepname, step := range ts.wf.Steps {
+			if !strings.HasPrefix(stepname, "wait-") {
+				continue
+			}
+			if err := ts.wf.AddDependency(copyStep, step); err != nil {
+				return fmt.Errorf("failed to add copy-objects step dependency to workflow %s: %v", ts.Name, err)
+			}
+		}
+
+	}
+	return nil
+}
+
+type testResult struct {
+	testWorkflow    *TestWorkflow
+	skipped         bool
+	workflowSuccess bool
+	err             error
+	results         []string
+}
+
+func getTestResults(ctx context.Context, ts *TestWorkflow) ([]string, error) {
+	results := []string{}
+	createVMsStep := ts.wf.Steps[createVMsStepName]
+	for _, vm := range createVMsStep.CreateInstances.Instances {
+		log.Printf("going to download %s", vm.Metadata["_test_results_url"])
+		out, err := utils.DownloadGCSObject(ctx, client, vm.Metadata["_test_results_url"])
+		if err != nil {
+			return nil, fmt.Errorf("failed to get results for test %s vm %s: %v", ts.Name, vm.Name, err)
+		}
+		results = append(results, string(out))
+	}
+
+	return results, nil
+}
+
+// NewTestWorkflow returns a new TestWorkflow.
+func NewTestWorkflow(name, image string) (*TestWorkflow, error) {
+	t := &TestWorkflow{}
+	t.Name = name
+	t.Image = image
+
+	parts := strings.Split(image, "/")
+	t.ShortImage = parts[len(parts)-1]
+
+	t.wf = daisy.New()
+	t.wf.Name = strings.ReplaceAll(name, "_", "-")
+
+	return t, nil
+}
+
+// PrintTests prints all test workflows.
+func PrintTests(ctx context.Context, testWorkflows []*TestWorkflow, project, zone string) {
+	finalizeWorkflows(testWorkflows, zone, project)
+	for _, test := range testWorkflows {
+		if test.wf == nil {
+			continue
+		}
+		test.wf.Print(ctx)
+	}
+}
+
+// ValidateTests validates all test workflows.
+func ValidateTests(ctx context.Context, testWorkflows []*TestWorkflow, project, zone string) error {
+	finalizeWorkflows(testWorkflows, zone, project)
+	for _, test := range testWorkflows {
+		if test.wf == nil {
+			continue
+		}
+		if err := test.wf.Validate(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RunTests runs all test workflows.
+func RunTests(ctx context.Context, testWorkflows []*TestWorkflow, outPath, project, zone string, parallelCount int) {
+	var err error
+	client, err = storage.NewClient(ctx)
+	if err != nil {
+		log.Fatalf("Failed to set up storage client: %v", err)
+	}
+	finalizeWorkflows(testWorkflows, zone, project)
+
+	testResults := make(chan testResult, len(testWorkflows))
+	testchan := make(chan *TestWorkflow, len(testWorkflows))
+
+	var wg sync.WaitGroup
+	for i := 0; i < parallelCount; i++ {
```

Comment:
if `parallelCount` is greater than `len(testWorkflows)` I think this will dead lock waiting at `range testchan`. Add a check or trim `parallelCount` to ensure this doesn't happen.

Reply:
let's say there are 5 workflows to run but we set parallel count to 30
we will create all 30 goroutines, which will wait on test workflows to run
we will load the 5 workflows to the test channel and close it
the first 5 goroutines will each grab one
the remaining 25 goroutines will then see the closed channel and exit
the first 5 will exit once done with their work